[analyzer] Fix crash in Stream checker when using void pointers

steakhal opened the pull request:
We can get zero type size (thus div by zero crash) if the region is for a 'void*' pointer.
In this patch, let's just override the void type with a char type to avoid the crash.

Fixes
https://github.com/llvm/llvm-project/pull/93408#issuecomment-2189766510

NagyDonat reviewed (changes requested):
Unfortunately this PR is not a full solution, because e.g. the following test code still triggers the crash (if it is appended to the test file `stream.c`):
```c
struct zerosized {
    int foo[0];
};

void fread_zerosized(struct zerosized *buffer) {
  FILE *f = fopen("/tmp/foo.txt", "r");
  fread(buffer, 1, 1, f);
  fclose(f);
}
```

(I know that zero-sized arrays are not standard-compliant, but they are a widespread extension: e.g. both clang and gcc accept this `struct zerosized` with the default settings.)

Overall, I'd say that it's futile to try to recognize zero-sized types with a "canonical type equal to" check, so you should just check whether `ElemSizeInChars` is zero and do something based on that. (Either an early return, or you can say `ElemSizeInChars = 1` at that point if you think that that's the logically correct solution.)

`<bikeshedding>`This way you could also avoid the immediately invoked lambda in `getPointeeType` which is really ugly in my opinion.`</bikeshedding>`

NagyDonat commented:
_I only noticed that this PR was already merged after posting the review. There is no need to revert the commit -- it's better than nothing -- but I'd be happy if you created a followup change that also handles the testcase that I mentioned._

steakhal commented:
> Overall, I'd say that it's futile to try to recognize zero-sized types with a "canonical type equal to" check, so you should just check whether `ElemSizeInChars` is zero and do something based on that. (Either an early return, or you can say `ElemSizeInChars = 1` at that point if you think that that's the logically correct solution.)

Yes :/

> `<bikeshedding>`This way you could also avoid the immediately invoked lambda in `getPointeeType` which is really ugly in my opinion.`</bikeshedding>`

Alright. Duplicated the call I wanted to avoid, but I agree it looks better now.

> _I only noticed that this PR was already merged after posting the review. There is no need to revert the commit -- it's better than nothing -- but I'd be happy if you created a followup change that also handles the testcase that I mentioned._

I'm not sure what you refer to. This PR is not approved, hence not merged.
Please continue with the review.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -1034,16 +1034,16 @@ void StreamChecker::preWrite(const FnDescription *Desc, const CallEvent &Call,
+static QualType getPointeeType(const MemRegion *R) {
   if (!R)
-    return std::nullopt;
+    return {};
```
It's a bit strange that here you return `{}`, while at the end of the function `QualType{}` is explicitly specified.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:
```diff
@@ -1034,16 +1034,16 @@ void StreamChecker::preWrite(const FnDescription *Desc, const CallEvent &Call,
+    return {};
   if (const auto *ER = dyn_cast<ElementRegion>(R))
-    return ER->getElementType();
+    return ER->getElementType()->getCanonicalTypeUnqualified();
```
Consider moving the `->getCanonicalTypeUnqualified()` calls into `escapeByStartIndexAndCount()` because that's the only place where it will be relevant. (This would eliminate the code duplication.)

steakhal replied:
Actually, now that I don't compare the type against a concrete alternative, it's no longer relevant. I'll just drop it. Thanks for spotting.

NagyDonat reviewed (approved):
Thanks for the updates! I added few minor comments, but the PR is already good enough to be merged.

> >  I only noticed that this PR was already merged after posting the review. There is no need to revert the commit -- it's better than nothing -- but I'd be happy if you created a followup change that also handles the testcase that I mentioned.
>
> I'm not sure what you refer to. This PR is not approved, hence not merged. Please continue with the review.

Oops, sorry -- I was probably confused by the purple "merged" icon in 
![image](https://github.com/llvm/llvm-project/assets/43410265/e4b494d4-014f-4b03-942b-00debb90a709)
...but the real reason is that I got up at 4 AM today :sleeping:

steakhal committed:
Replace immediately invoked lambda by duplicating code

steakhal committed:
Harmonize return paths

steakhal committed:
Prefer directly checking against 0

steakhal committed:
[analyzer] Fix crash in Stream checker when using void pointers

We can get zero type size (thus div by zero crash) if the region is for
a 'void*' pointer.
In this patch, let's just override the void type with a char type to
avoid the crash.

Fixes
https://github.com/llvm/llvm-project/pull/93408#issuecomment-2189766510

steakhal committed:
Drop the getCanonicalTypeUnqualified calls
